File: backend/internal/domain/facility/bacnet_object.go
package facility

import (
	"github.com/besart951/go_infra_link/backend/internal/domain"
	"github.com/google/uuid"
)

type BacnetObject struct {
	domain.Base
	TextFix        string `gorm:"index:idx_field_device_textfix;not null"`
	Description    *string
	GMSVisible     bool `gorm:"default:false"`
	Optional       bool `gorm:"default:false"`
	TextIndividual *string

	SoftwareType   BacnetSoftwareType `gorm:"type:varchar(50);not null"`
	SoftwareNumber uint16             `gorm:"not null"`

	HardwareType     BacnetHardwareType `gorm:"type:varchar(50)"`
	HardwareQuantity uint8

	FieldDeviceID       *uuid.UUID         `gorm:"type:uuid;index;index:idx_field_device_textfix"`
	FieldDevice         *FieldDevice       `gorm:"foreignKey:FieldDeviceID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	SoftwareReferenceID *uuid.UUID         `gorm:"type:uuid;index"`
	SoftwareReference   *BacnetObject      `gorm:"foreignKey:SoftwareReferenceID"`
	StateTextID         *uuid.UUID         `gorm:"type:uuid;index"`
	StateText           *StateText         `gorm:"foreignKey:StateTextID"`
	NotificationClassID *uuid.UUID         `gorm:"type:uuid;index"`
	NotificationClass   *NotificationClass `gorm:"foreignKey:NotificationClassID"`
	AlarmTypeID         *uuid.UUID         `gorm:"type:uuid;index"`
	AlarmType           *AlarmType         `gorm:"foreignKey:AlarmTypeID"`
	AlarmDefinitionID   *uuid.UUID         `gorm:"-:all"`
}

// BacnetObjectPatch represents a partial update for a bacnet object.
// Only non-nil fields are applied.
type BacnetObjectPatch struct {
	ID                  uuid.UUID
	TextFix             *string
	Description         *string
	GMSVisible          *bool
	Optional            *bool
	TextIndividual      *string
	SoftwareType        *BacnetSoftwareType
	SoftwareNumber      *uint16
	HardwareType        *BacnetHardwareType
	HardwareQuantity    *uint8
	SoftwareReferenceID *uuid.UUID
	StateTextID         *uuid.UUID
	NotificationClassID *uuid.UUID
	AlarmTypeID         *uuid.UUID
	AlarmDefinitionID   *uuid.UUID
}

// ApplyPatch applies the supported partial state transition to an existing
// BACnet object. Persistence, parent existence, and cross-object uniqueness
// remain application/service concerns.
func (b *BacnetObject) ApplyPatch(patch BacnetObjectPatch) error {
	if b == nil || b.ID == uuid.Nil {
		return domain.ErrInvalidArgument
	}
	if patch.ID != uuid.Nil && patch.ID != b.ID {
		return domain.ErrInvalidArgument
	}

	if patch.TextFix != nil {
		b.TextFix = *patch.TextFix
	}
	if patch.Description != nil {
		b.Description = cloneBacnetPointer(patch.Description)
	}
	if patch.GMSVisible != nil {
		b.GMSVisible = *patch.GMSVisible
	}
	if patch.Optional != nil {
		b.Optional = *patch.Optional
	}
	if patch.TextIndividual != nil {
		b.TextIndividual = cloneBacnetPointer(patch.TextIndividual)
	}
	if patch.SoftwareType != nil {
		b.SoftwareType = *patch.SoftwareType
	}
	if patch.SoftwareNumber != nil {
		b.SoftwareNumber = *patch.SoftwareNumber
	}
	if patch.HardwareType != nil {
		b.HardwareType = *patch.HardwareType
	}
	if patch.HardwareQuantity != nil {
		b.HardwareQuantity = *patch.HardwareQuantity
	}
	if patch.SoftwareReferenceID != nil {
		b.SoftwareReferenceID = cloneBacnetPointer(patch.SoftwareReferenceID)
	}
	if patch.StateTextID != nil {
		b.StateTextID = cloneBacnetPointer(patch.StateTextID)
	}
	if patch.NotificationClassID != nil {
		b.NotificationClassID = cloneBacnetPointer(patch.NotificationClassID)
	}
	if patch.AlarmTypeID != nil {
		b.AlarmTypeID = cloneBacnetPointer(patch.AlarmTypeID)
	}
	if patch.AlarmDefinitionID != nil {
		b.AlarmDefinitionID = cloneBacnetPointer(patch.AlarmDefinitionID)
	}

	return nil
}

// AssignToFieldDevice changes the direct facility owner. Project-link
// reconciliation and uniqueness validation are deliberately outside the
// entity because they require repositories.
func (b *BacnetObject) AssignToFieldDevice(fieldDeviceID uuid.UUID) error {
	if b == nil || b.ID == uuid.Nil || fieldDeviceID == uuid.Nil {
		return domain.ErrInvalidArgument
	}
	b.FieldDeviceID = &fieldDeviceID
	return nil
}

// DetachFromFieldDevice prepares a BACnet object for indirect ownership, such
// as an ObjectData template association managed by the application service.
func (b *BacnetObject) DetachFromFieldDevice() error {
	if b == nil || b.ID == uuid.Nil {
		return domain.ErrInvalidArgument
	}
	b.FieldDeviceID = nil
	return nil
}

func cloneBacnetPointer[T any](value *T) *T {
	if value == nil {
		return nil
	}
	clone := *value
	return &clone
}

type BacnetSoftwareType string

const (
	BacnetSoftwareTypeAI BacnetSoftwareType = "ai"
	BacnetSoftwareTypeAO BacnetSoftwareType = "ao"
	BacnetSoftwareTypeAV BacnetSoftwareType = "av"
	BacnetSoftwareTypeBI BacnetSoftwareType = "bi"
	BacnetSoftwareTypeBO BacnetSoftwareType = "bo"
	BacnetSoftwareTypeBV BacnetSoftwareType = "bv"
	BacnetSoftwareTypeMI BacnetSoftwareType = "mi"
	BacnetSoftwareTypeMO BacnetSoftwareType = "mo"
	BacnetSoftwareTypeMV BacnetSoftwareType = "mv"
	BacnetSoftwareTypeCA BacnetSoftwareType = "ca"
	BacnetSoftwareTypeEE BacnetSoftwareType = "ee"
	BacnetSoftwareTypeLP BacnetSoftwareType = "lp"
	BacnetSoftwareTypeNC BacnetSoftwareType = "nc"
	BacnetSoftwareTypeSC BacnetSoftwareType = "sc"
	BacnetSoftwareTypeTL BacnetSoftwareType = "tl"
)

type BacnetHardwareType string

const (
	BacnetHardwareTypeEMPTY BacnetHardwareType = ""
	BacnetHardwareTypeDO    BacnetHardwareType = "do"
	BacnetHardwareTypeAO    BacnetHardwareType = "ao"
	BacnetHardwareTypeDI    BacnetHardwareType = "di"
	BacnetHardwareTypeAI    BacnetHardwareType = "ai"
)

File: frontend/src/lib/domain/facility/bacnet-object.ts
/**
 * BACnet Object domain types
 * Mirrors backend: internal/domain/facility/bacnet_object.go
 */

export interface BacnetObject {
	id: string;
	text_fix: string;
	description?: string;
	gms_visible: boolean;
	optional: boolean;
	text_individual?: string;
	software_type: string;
	software_number: number;
	hardware_type: string;
	hardware_quantity: number;
	field_device_id?: string;
	software_reference_id?: string;
	state_text_id?: string;
	notification_class_id?: string;
	alarm_definition_id?: string;
	created_at: string;
	updated_at: string;
}

export interface CreateBacnetObjectRequest {
	text_fix: string;
	description?: string;
	gms_visible: boolean;
	optional: boolean;
	text_individual?: string;
	software_type: string;
	software_number: number;
	hardware_type: string;
	hardware_quantity: number;
	field_device_id?: string;
	software_reference_id?: string;
	state_text_id?: string;
	notification_class_id?: string;
	alarm_definition_id?: string;
}

export interface UpdateBacnetObjectRequest {
	text_fix?: string;
	description?: string;
	gms_visible?: boolean;
	optional?: boolean;
	text_individual?: string;
	software_type?: string;
	software_number?: number;
	hardware_type?: string;
	hardware_quantity?: number;
	field_device_id?: string;
	software_reference_id?: string;
	state_text_id?: string;
	notification_class_id?: string;
	alarm_definition_id?: string;
}

// BACnet Software Types
export const BACNET_SOFTWARE_TYPES = [
	{ value: 'ai', label: 'AI - Analog Input' },
	{ value: 'ao', label: 'AO - Analog Output' },
	{ value: 'av', label: 'AV - Analog Value' },
	{ value: 'bi', label: 'BI - Binary Input' },
	{ value: 'bo', label: 'BO - Binary Output' },
	{ value: 'bv', label: 'BV - Binary Value' },
	{ value: 'mi', label: 'MI - Multi-state Input' },
	{ value: 'mo', label: 'MO - Multi-state Output' },
	{ value: 'mv', label: 'MV - Multi-state Value' },
	{ value: 'ca', label: 'CA - Calendar' },
	{ value: 'ee', label: 'EE - Event Enrollment' },
	{ value: 'lp', label: 'LP - Life Safety Point' },
	{ value: 'nc', label: 'NC - Notification Class' },
	{ value: 'sc', label: 'SC - Schedule' },
	{ value: 'tl', label: 'TL - Trend Log' }
] as const;

// BACnet Hardware Types
export const BACNET_HARDWARE_TYPES = [
	{ value: 'do', label: 'DO - Digital Output' },
	{ value: 'ao', label: 'AO - Analog Output' },
	{ value: 'di', label: 'DI - Digital Input' },
	{ value: 'ai', label: 'AI - Analog Input' }
] as const;
